includes/razorpay-subscriptions.php

```diff
@@ -0,0 +1,246 @@
+
+        return $subscription['id'];
+    }
+
+    public function cancelSubscription($subscriptionId)
+    {
+        $subscription = $this->api->subscription->cancel($subscriptionId);
+    }
+
+    public function getProduct($order)
+    {
+        $products = $order->get_items();
+
+        // Technically, subscriptions work only if there's one array in the cart
+        if (sizeof($products) > 1)
+        {
+            throw new Exception('Currently Razorpay does not support more than
+                                one product in the cart if one of the products
+                                is a subscription.');
+        }
+
+        return array_values($products)[0];
+    }
+
+    protected function getSubscriptionCreateData($orderId)
+    {
+        $order = new WC_Order($orderId);
+
+        $product = $this->getProduct($order);
+
+        $planId = $this->getProductPlanId($product);
+
+        $customerId = $this->getCustomerId($order);
+
+        $length = (int) WC_Subscriptions_Product::get_length($product['product_id']);
+
+        $subscriptionData = array(
+            'customer_id'     => $customerId,
+            'plan_id'         => $planId,
+            'quantity'        => (int) $product['qty'],
+            'total_count'     => $length,
+            'customer_notify' => 0,
+            'notes'           => array(
+                'woocommerce_order_id'   => $orderId,
+                'woocommerce_product_id' => $product['product_id']
+            ),
+        );
+
+        $signUpFee = WC_Subscriptions_Product::get_sign_up_fee($product['product_id']);
+
+        if ($signUpFee)
+        {
+            $subscriptionData['addons'] = array(
+                array (
+                    'item' => array(
+                        'amount'   => (int) round($signUpFee * 100),
+                        'currency' => get_woocommerce_currency(),
+                        'name'     => $product['name']
+                    )
+                )
+            );
+        }
+
+        return $subscriptionData;
+    }
+
+    protected function getProductPlanId($product)
+    {
+        $productId = $product['product_id'];
+
+        $metadata = get_post_meta($productId);
+
+        // Creating a new plan only if plan isn't created already
+        // Or if the plan has changed
+        if ($this->shouldCreatePlan($metadata, $productId) === true)
+        {
+            $planId = $this->createPlan($product);
+
+            add_post_meta($productId, 'razorpay_wc_plan_id', $planId, true);
+        }
+        else
+        {
+            // Retrieve the plan id if already created
+            $planId = $metadata['razorpay_wc_plan_id'][0];
+        }
+
+        return $planId;
+    }
+
+    /**
+     * We shouldn't create plan only if the following conditions are met
+     * 1. There exists a plan id as product metadata
+     * 2. The corresponding plan has amount a1
+     * 3. The sale price is amount a2
```

Comment:
What if the get_price has changed to a3, wouldn't that require us to create a new plan as well?